File: 1584.min-cost-to-connect-all-points.go
/*
 * @lc app=leetcode id=1584 lang=golang
 *
 * [1584] Min Cost to Connect All Points
 */

package main

import (
	"container/heap"
	"math"
)

// @lc code=start
type Item struct {
	distance, point int
}

//	type Item struct {
//		distance, point int
//	}
type PriorityQueue1584 []Item

func (pq PriorityQueue1584) Len() int { return len(pq) }

func (pq PriorityQueue1584) Less(i, j int) bool {
	return pq[i].distance < pq[j].distance
}

func (pq PriorityQueue1584) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PriorityQueue1584) Push(x interface{}) {
	item := x.(Item)
	*pq = append(*pq, item)
}

func (pq *PriorityQueue1584) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

func manhattanDistance(p1, p2 []int) int {
	return int(math.Abs(float64(p1[0]-p2[0])) + math.Abs(float64(p1[1]-p2[1])))
}

//	func minCostConnectPoints(points [][]int) int {
//		heapDict := make(map[int]int)
//		for i := range points {
//			heapDict[i] = int(math.MaxInt64) // Initialize all distances to infinity
//		}
//		heapDict[0] = 0 // Start node
//
//		queue := make(PriorityQueue, 1)
//		queue[0] = Item{distance: 0, point: 0}
//		heap.Init(&queue)
//
//		visited := make(map[int]bool, len(points))
//		mstWeight := 0
//
//		for queue.Len() > 0 {
//			current := heap.Pop(&queue).(Item)
//			weight, point := current.distance, current.point
//
//			if visited[point] || heapDict[point] < weight {
//				continue
//			}
//
//			visited[point] = true
//			mstWeight += weight
//
//			for v := range points {
//				if !visited[v] {
//
//					newDistance := manhattanDistance(points[point], points[v])
//
//					if newDistance < heapDict[v] {
//						heapDict[v] = newDistance
//						heap.Push(&queue, Item{distance: newDistance, point: v})
//					}
//				}
//			}
//		}
//
//		return mstWeight
//	}
func minCostConnectPoints(points [][]int) int {
	minDist := make(map[int]int)
	for i := range points {
		minDist[i] = int(math.MaxInt64) // Initialize all distances to infinity
	}
	minDist[0] = 0 // Start node

	queue := make(PriorityQueue1584, 1)
	queue[0] = Item{distance: 0, point: 0}
	heap.Init(&queue)

	visited := make(map[int]bool, len(points))
	sum := 0

	for queue.Len() > 0 {
		current := heap.Pop(&queue).(Item)
		currDist, point := current.distance, current.point

		if visited[point] || minDist[point] < currDist {
			continue
		}

		visited[point] = true
		sum += currDist

		for v := range points {
			if !visited[v] {
				newDistance := manhattanDistance(points[point], points[v])

				if newDistance < minDist[v] {
					minDist[v] = newDistance
					heap.Push(&queue, Item{distance: newDistance, point: v})
				}
			}
		}
	}

	return sum
}

// @lc code=end

File: 1584.min-cost-to-connect-all-points.ts
/*
 * @lc app=leetcode id=1584 lang=typescript
 *
 * [1584] Min Cost to Connect All Points
 */

// @lc code=start
function minCostConnectPoints(points: number[][]): number {
  function prioPop(queue: number[]) {
    let minIndex = 0;
    let minDist = Infinity;

    for (let i = 0; i < queue.length; i++) {
      if (minDist > distance[queue[i]]) {
        minIndex = i;
        minDist = distance[queue[i]];
      }
    }
    let tmp = queue[minIndex];
    queue[minIndex] = queue[queue.length - 1];
    queue[queue.length - 1] = tmp;

    return queue.pop();
  }
  const len = points.length;
  let visited = Array(len).fill(false);
  let visitsCount = 0;
  let weight = 0;
  const queue = [0];

  const distance = Array(len).fill(Infinity);
  distance[0] = 0;

  while (queue.length) {
    let current = prioPop(queue);
    let currentDist = distance[current];
    if (visited[current]) {
      continue;
    }
    visited[current] = true;
    visitsCount++;
    weight += currentDist;

    for (let i = 0; i < len; i++) {
      if (visited[i]) {
        continue;
      }

      let manhattanDist =
        Math.abs(points[current][0] - points[i][0]) +
        Math.abs(points[current][1] - points[i][1]);
      if (manhattanDist < distance[i]) {
        distance[i] = manhattanDist;
        queue.push(i);
      }
    }
  }

  return weight;
}

// @lc code=end
